### PR TITLE
Context extensions

### LINKED ISSUE #4445: Add NonSerializable Context entries
Currently all context parameters are serialized to JSON when stored. This means that they can be sent to coprocessors and Rhai.

However this is a not possible for certain structures, and is at the very least could affect performance for certain scenarios.

Apollo should continue to use the existing json based context where appropriate, but also provide functionality similar to `private_entries` for users.

As negative trait bounds don't exist yet we will need to add a new API for this. Suggest that we model using by using the existing API with a suffix.

### PR DESCRIPTION
**Pull request by BrynCooke**
Plugin authors may need to store items in context that are not serializable. This could be something like an HTTP client, or something else that cannot be serialized to a Json object.

This PR renames the existing `private_entries` in context to `extensions` and adds an accessor.
Care should be take to not hold onto mutex guards for too long, however this is a non-issue because:
* Users should not be performing blocking operations from within non async blocks. This is not a new issue and we can't prevent them from doing this.
* MutexGuard is not send, so users won't be able to hold the lock across async awaits anyway.

Fixes #4445

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Review by Geal — APPROVED** [review]

**Review by bnjjj — APPROVED** [review]

**Comment by Geal** [comment]
we won't be able to prevent users from locking a mutex for too long, but we could at least make sure that in router internal code, we're not blocked by this and have failure modes. One idea would be to use [try_lock_until](https://docs.rs/lock_api/0.4.9/lock_api/struct.Mutex.html#method.try_lock_until) and have a way to shutdown request processing if we reach the timeout

**Comment by BrynCooke** [comment]
I added some code to detect long holds of context.
This introduces a new sync Extensions struct, so please take a look at the API again.

I also pub use Extensions from the root package. There's no going back with this, so make sure you are OK with this.

@SimonSapin Would like you also have your approval before merging.

File: apollo-router/src/context/extensions/sync.rs
Comment (BrynCooke) [reply apollo-router/src/context/extensions/sync.rs:62]: It's because a bunch of unit tests fail in single threaded mode as they are blocking context for significant amounts of time. 
These could actually be bugs, but we'll need to follow up to investigate.
I'll make the comment clearer.

Comment (BrynCooke) [inline apollo-router/src/context/extensions/sync.rs:62]:
```suggestion
        // In debug builds we check that extensions is never held for too long.
        // We  only check if the current runtime is multi-threaded, because a bunch of unit tests fail the assertion and these need to be investigated separately. 
```

File: apollo-router/src/context/extensions/sync.rs
Comment (BrynCooke) [reply apollo-router/src/context/extensions/sync.rs]: [4af0d2f](https://github.com/apollographql/router/pull/4474/commits/4af0d2f2aaf0f3e723e8de1cacd1bcf3e21d582f)

File: apollo-router/src/lib.rs
Comment (BrynCooke) [reply apollo-router/src/lib.rs]: [4af0d2f](https://github.com/apollographql/router/pull/4474/commits/4af0d2f2aaf0f3e723e8de1cacd1bcf3e21d582f)

**Review by SimonSapin — APPROVED** [review]

**Pull request merged**
